### PR TITLE
Half polarized workflow

### LINKED ISSUE #61: Add workflow for half-polarized data

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #61.

### CODE REVIEW

File: docs/user-guide/workflow.ipynb [hunk 1]
```diff
@@ -139,6 +139,36 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from ess.polarization.correction import CorrectionWorkflow\n",
+    "\n",
+    "results = (\n",
+    "    pol.HalfPolarizedCorrectedData[pol.Up],\n",
+    "    pol.HalfPolarizedCorrectedData[pol.Down],\n",
+    ")\n",
```
Comment (astellhorn) [inline docs/user-guide/workflow.ipynb:161]: Here, `pol.Up` and `pol.Down` stands for the polarizer spin, or for the "dummy analyzer spin"?
As they are part of `pol.HalfPolarizedCorrectedData` ( in correction.py) and defined via `up=p_up` and `down=p_down`, its the polarizer I think? I was just wondering about where at all the "dummy value for the spin of the analyzer that is not present." (code line 170) is used.

File: docs/user-guide/workflow.ipynb [hunk 2]
```diff
@@ -139,6 +139,36 @@
...
+    "workflow = CorrectionWorkflow(half_polarized=True)\n",
+    "workflow.visualize(results, graph_attr={\"rankdir\": \"LR\"})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that in parts of the workflow `Up` is used as a dummy value for the spin of the analyzer that is not present."
+   ]
```
Comment (astellhorn) [inline docs/user-guide/workflow.ipynb:171]: see above

File: docs/user-guide/workflow.ipynb (hunk 1 above)
Comment (SimonHeybrock) [reply docs/user-guide/workflow.ipynb:161]: What you see here is *not* the dummy value. The dummy value is just internal, so we can reuse the workflow graph (and I think you need it when setting the input data).

File: tests/correction_test.py [hunk 3]
```diff
@@ -240,3 +241,26 @@ def test_polarization_analysis_workflow_creation(
...
+    ground_truth = np.array([7.0, 11.0])
+    polarizer = np.array([[1.1, 0.7], [0.7, 1.1]])
+    flipper = np.array([[1.0, 0.0], [1 - f, f]])
+    intensity = flipper @ polarizer @ ground_truth
+
+    workflow = CorrectionWorkflow(half_polarized=True)
+    workflow[TransmissionFunction[Polarizer]] = FakeTransmissionFunction(polarizer)
+    workflow[ReducedSampleDataBySpinChannel[Up, pol.NoAnalyzer]] = intensity[0]
+    workflow[ReducedSampleDataBySpinChannel[Down, pol.NoAnalyzer]] = intensity[1]
+    workflow[FlipperEfficiency[Polarizer]] = FlipperEfficiency(f)
```
Comment (SimonHeybrock) [inline tests/correction_test.py:259]: @astellhorn Have a look at these lines, they should answer your comments above: Here I used `NoAnalyzer` as a value for the "analyzer spin", but this is just an internal alias for "Up" (arbitrary choice).

File: src/ess/polarization/types.py [hunk 4]
```diff
@@ -11,6 +11,8 @@
 Down = NewType('Down', int)
 PolarizerSpin = TypeVar('PolarizerSpin', Up, Down)
 AnalyzerSpin = TypeVar('AnalyzerSpin', Up, Down)
+NoAnalyzer = Up
+"""Indicates that element is not in use, e.g., analyzer in half-polarized case."""
```
Comment (SimonHeybrock) [inline src/ess/polarization/types.py:15]: @astellhorn See also here.

File: docs/user-guide/workflow.ipynb (hunk 1 above)
Comment (astellhorn) [reply docs/user-guide/workflow.ipynb:161]: Ok!

File: tests/correction_test.py (hunk 3 above)
Comment (astellhorn) [reply tests/correction_test.py:259]: Yes I understand

File: src/ess/polarization/correction.py [hunk 5]
```diff
@@ -208,3 +234,27 @@ def PolarizationAnalysisWorkflow(
         TransmissionFunction[Analyzer]
     ]
     return workflow
+
+
+def HalfPolarizedAnalysisWorkflow(
```
Comment (astellhorn) [inline src/ess/polarization/correction.py:239]: If this is not too big of a change, I recommend to call this `HalfPolarizationWorkflow`. (As long as it is defined clearly this name does not really matter, but most people understand the "FullPolarization" - with Analyzer - when "Analysis" is mentioned.)

File: src/ess/polarization/correction.py [hunk 6]
```diff
@@ -208,3 +234,27 @@ def PolarizationAnalysisWorkflow(
...
+    *,
+    polarizer_workflow: sciline.Pipeline,
+) -> sciline.Pipeline:
+    """
+    Create a half-polarized analysis workflow, i.e, with a polarizer but no analyzer.
```
Comment (astellhorn) [inline src/ess/polarization/correction.py:244]: Equivalent, for the documentary for later/for users I write "Create a half-polarized workflow"...

File: tests/correction_test.py (hunk 3 above)
Comment (astellhorn) [reply tests/correction_test.py:259]: For testing the polarizer: useful values to test would be either (i)  for a supermirror T_Up + T_Down = 1 (e.g. 0.96 and 0.04), or for (ii) a 3He cell T_Up + T_Down < 1 (1 only for perfect transmisison through the cell), (e.g. 0.8 and 0.05).

Comment (astellhorn) [reply tests/correction_test.py:259]: I am sorry I still somehow cant test this myself as when compiling the `correction_test.py`, I still get the error message `cannot import name 'assert_allclose' from 'scipp.testing'`

Comment (SimonHeybrock) [reply tests/correction_test.py:259]: You may need to update your Scipp version.

**Comment by jokasimr** [comment]
> I am sorry I still somehow cant test this myself as when compiling the correction_test.py, I still get the error message cannot import name 'assert_allclose' from 'scipp.testing'

Sounds like you might have an old scipp version installed. Try to update it and see if the issue goes away.
If you installed scipp using `mamba` or `conda` do:
```
mamba update -c scipp scipp
```
If you installed using `pip` do
```
pip install scipp --upgrade
```

**Comment by astellhorn** [comment]
> pip install scipp --upgrade

Yes, it works! Thanks!

File: tests/correction_test.py (hunk 3 above)
Comment (SimonHeybrock) [reply tests/correction_test.py:259]: > For testing the polarizer: useful values to test would be either (i) for a supermirror T_Up + T_Down = 1 (e.g. 0.96 and 0.04), or for (ii) a 3He cell T_Up + T_Down < 1 (1 only for perfect transmisison through the cell), (e.g. 0.8 and 0.05).

@astellhorn Can you clarify which values you mean? `f`? It is being tested for a range of values already.

Comment (astellhorn) [reply tests/correction_test.py:259]: I meant the polarizer matrix values. Currently they are `polarizer = np.array([[1.1, 0.7], [0.7, 1.1]])`.

Comment (SimonHeybrock) [reply tests/correction_test.py:259]: Ok, I updated it. IT does not actually matter for this test, as long as the matrix structure is right and it has an inverse.

Comment (astellhorn) [reply tests/correction_test.py:259]: Thanks! Yes, that is true. It was more for clarity.

**Pull request merged**